### PR TITLE
Grpc hot reload

### PR DESCRIPTION
**Pull request by WilfredAlmeida**

### CODE REVIEW

File: client/main.rs [hunk 1]
```diff
@@ -60,8 +70,18 @@ struct Params {
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn std::error::Error>> {
     env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
+
     let params = Params::from_args();
 
+    let (grpc_urls_channel_sender, mut grpc_urls_channel_receiver) =
+        tokio::sync::mpsc::channel::<Vec<String>>(1);
+
+    let mut grpc_urls_from_file = read_grpc_urls_from_file(params.grpc_urls_file).unwrap();
```
Comment (fanatid) [inline client/main.rs]: better return error than making `unwrap`

File: client/main.rs [hunk 2]
```diff
@@ -80,29 +100,64 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+    loop {
+        match grpc_urls_channel_receiver.recv().await {
+            Some(new_grpc_urls) => {
+                let urls_to_spawn: Vec<String> = new_grpc_urls
+                    .iter()
+                    .filter(|x| !grpc_urls_from_file.contains(x))
+                    .map(|x| x.to_string())
+                    .collect();
+
+                let urls_to_abort: Vec<String> = grpc_urls_from_file
```
Comment (fanatid) [inline client/main.rs]: no need to collect to `Vec` because you consume it later

File: client/main.rs [hunk 3]
```diff
@@ -80,29 +100,64 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                    .iter()
+                    .filter(|x| !new_grpc_urls.contains(x))
+                    .map(|x| x.to_string())
+                    .collect();
+
+                for i in urls_to_abort {
+                    let mut all_tasks = all_tasks.write().await;
```
Comment (fanatid) [inline client/main.rs]: it doesn't required to acquire write lock for every abort, better acquire once for all aborts

File: client/main.rs [hunk 4]
```diff
@@ -80,29 +100,64 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                    if let Some(task) = all_tasks.remove(&i) {
+                        info!("Aborting task for url: {:?}", i);
+                        task.abort();
+                    }
+                }
+
+                grpc_urls_from_file.clear();
+                grpc_urls_from_file.extend(new_grpc_urls.clone());
+
+                if urls_to_spawn.len() == 0 {
```
Comment (fanatid) [inline client/main.rs]:
```suggestion
                if urls_to_spawn.is_empty() {
```

File: client/main.rs [hunk 5]
```diff
@@ -80,29 +100,64 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                    continue;
+                }
+
+                info!("Spawning tasks for new urls: {:?}", urls_to_spawn);
+
+                build_tasks(
+                    urls_to_spawn,
+                    &Params::from_args(),
+                    tx_transactions.clone(),
+                    Arc::clone(&all_tasks),
```
Comment (fanatid) [inline client/main.rs]: `Arc` should not be required here, we do not spawn `build_tasks`

File: client/main.rs [hunk 6]
```diff
@@ -142,3 +197,100 @@ async fn spawn_grpc_connection_with_retry(
...
+    ));
+}
+
+async fn build_tasks(
+    grpc_urls: Vec<String>,
+    params: &Params,
+    tx_transactions: UnboundedSender<ForwardedTransaction>,
+    all_tasks: Arc<RwLock<HashMap<String, JoinHandle<()>>>>,
+) {
+    for i in grpc_urls.clone() {
```
Comment (fanatid) [inline client/main.rs]:
```suggestion
    for i in grpc_urls {
```

File: client/main.rs [hunk 7]
```diff
@@ -142,3 +197,100 @@ async fn spawn_grpc_connection_with_retry(
...
+        let mut all_tasks = all_tasks.write().await;
```
Comment (fanatid) [inline client/main.rs]: need to be moved outside of the `for-in` loop

File: client/main.rs [hunk 8]
```diff
@@ -142,3 +197,100 @@ async fn spawn_grpc_connection_with_retry(
...
+
+        let grpc_parsed_url: Uri = i.parse().unwrap();
```
Comment (fanatid) [inline client/main.rs]: no `.unwrap()` please, log the error

File: client/main.rs [hunk 9]
```diff
@@ -80,29 +100,64 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                for i in urls_to_abort {
+                    let mut all_tasks = all_tasks.write().await;
+                    if let Some(task) = all_tasks.remove(&i) {
+                        info!("Aborting task for url: {:?}", i);
+                        task.abort();
+                    }
+                }
+
+                grpc_urls_from_file.clear();
+                grpc_urls_from_file.extend(new_grpc_urls.clone());
```
Comment (fanatid) [inline client/main.rs:134]:
```suggestion
                grpc_urls_from_file = new_grpc_urls;
```

File: client/main.rs (hunk 1 above)
Comment (WilfredAlmeida) [reply client/main.rs]: If this fine?
```
    let mut grpc_urls_from_file;

    if let Ok(urls) = read_grpc_urls_from_file(params.grpc_urls_file) {
        grpc_urls_from_file = urls;
    } else {
        panic!("Failed to read grpc urls from file")
    }
```

Comment (fanatid) [reply client/main.rs]: no need to panic about such error
- replace return type to `anyhow::Result<()>`
- change code to ```use anyhow::Context; read_grpc_urls_from_file(params.grpc_urls_file).context("failed to read gRPC urls from file")?;```

File: client/main.rs (hunk 2 above)
Comment (WilfredAlmeida) [reply client/main.rs]: Removing the type annotation errors
 ```type inside `async` block must be known in this context
cannot infer type```

Comment (fanatid) [reply client/main.rs]:
```rust
                let mut locked_all_tasks = all_tasks.write().await;
                for grpc in grpc_urls_from_file.iter().filter(|x| !new_grpc_urls.contains(x)) {
                    if let Some(task) = all_tasks.remove(grpc) {
                        info!("Aborting task for url: {grpc:?}");
                        task.abort();
                    }
                }
                drop(locked_all_tasks);
```

File: client/main.rs (hunk 9 above)
Comment (WilfredAlmeida) [reply client/main.rs:134]: I tried this, and, if `grpc_urls_from_file` had some values which got removed, they'd still remain in the vector. It's weird but I did what you suggested and when it didn't work, did the current one.

Comment (fanatid) [reply client/main.rs:134]: you also can try `std::mem::replace`

**Comment by fanatid** [comment]
ok, I see `.expect` now, it's better than `.unwrap()` because adding an extra message but producing an error message instead of panic is better

lgtm

**Review by linuskendall — APPROVED** [review]

**Pull request merged**